### PR TITLE
[SYCLomatic] Refine migration of 3D memory copy.

### PR DESCRIPTION
**Pull request by tangjj11**
Signed-off-by: Tang, Jiajun jiajun.tang@intel.com

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dpct/bindless_images.hpp [hunk 1]
```diff
@@ -744,6 +789,25 @@ static inline void dpct_memcpy(image_mem_wrapper *src, size_t w_offset_src,
   sycl::free(temp, q);
 }
 
+/// TODO:
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/bindless_images.hpp]: content of TODO?

**Review by zhiweij1 — APPROVED** [review]

File: clang/runtime/dpct-rt/include/dpct/memory.hpp [hunk 2]
```diff
@@ -77,6 +77,127 @@ class pitched_data {
...
+                               sycl::queue q);
+} // namespace detail
+} // namespace experimental
+#endif
+class image_matrix;
+namespace detail {
+static pitched_data to_pitched_data(image_matrix *image);
+}
+
+class mem_cpy_param_wrapper {
```
Comment (tangjj11) [inline clang/runtime/dpct-rt/include/dpct/memory.hpp]: Need add comment.

**Review by ziranzha — APPROVED** [review]

File: clang/runtime/dpct-rt/include/dpct/memory.hpp [hunk 3]
```diff
@@ -624,6 +661,44 @@ dpct_memcpy(sycl::queue &q, void *to_ptr, const void *from_ptr,
...
+dpct_memcpy(sycl::queue &q, const memcpy_parameter &param) {
+  auto to = param.to.pitched;
+  auto from = param.from.pitched;
+#ifdef SYCL_EXT_ONEAPI_BINDLESS_IMAGES
+  if (param.to.image_bindless != nullptr &&
+      param.from.image_bindless != nullptr) {
+    std::vector<sycl::event> event_list;
+    host_buffer buf(param.size.size(), q, event_list);
+    to.set_data_ptr(buf.get_ptr());
+    // TODO: Need change logic when sycl support image_mem to image_mem copy.
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/memory.hpp]: PUt todo in L671.

File: clang/runtime/dpct-rt/include/dpct/memory.hpp [hunk 4]
```diff
@@ -77,6 +77,43 @@ class pitched_data {
...
+} // namespace experimental
+#endif
+class image_matrix;
+namespace detail {
+static pitched_data to_pitched_data(image_matrix *image);
+}
+
+/// Memory copy parameters for 2D/3D memory data.
+struct memcpy_parameter {
+  struct data_wrapper {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/memory.hpp:103]: for image_bindless/image,  who will init them to zero?

File: clang/runtime/dpct-rt/include/dpct/bindless_images.hpp (hunk 1 above)
Comment (tangjj11) [reply clang/runtime/dpct-rt/include/dpct/bindless_images.hpp]: fixed

File: clang/runtime/dpct-rt/include/dpct/memory.hpp (hunk 3 above)
Comment (tangjj11) [reply clang/runtime/dpct-rt/include/dpct/memory.hpp]: done

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**